src/relay/api.rs

```diff
@@ -274,12 +291,24 @@ async fn serve_websocket(
                         let envelope =
                             Envelope::parse_from_bytes(&msg).context("failed to load input message")?;
 
-                        // TODO: federation
                         //  instead of sending the message directly to the switch
                         //  we check federation information attached to the envelope
                         //  if federation is not empty and does not match the domain
                         //  of this server, we need to schedule this message to be
                         //  federated to the right relay (according to specs)
+                        match &envelope.federation {
```

Comment:
first of all, once message is sent to federation, you should do a `continue` because you don't need to do internal switching here. the message should then not be sent to the local switch

Also I think it's better syntax to do 
```rust
if matches!(envelope.federation, Some(fed) if fed != domain)) {
   /// federation
   continue;
}
```